### PR TITLE
[CloneFunction][DebugInfo] Avoid cloning DILocalVariables of inlined functions

### PR DESCRIPTION
**Pull request by dzhidzhoev**
- [DebugMetadata][DwarfDebug] Support function-local types in lexical block scopes (4/7)
- [CloneFunction][DebugInfo] Avoid cloning DILocalVariables of inlined functions

This is a follow-up for https://reviews.llvm.org/D144006, fixing a crash reported
in Chromium (https://reviews.llvm.org/D144006#4651955).

The first commit is added for convenience, as it has already been accepted.

If DISubpogram was not cloned (e.g. we are cloning a function that has other
functions inlined into it, and subprograms of the inlined functions are
not supposed to be cloned), it doesn't make sense to clone its
DILocalVariables as well.
Otherwise get duplicated DILocalVariables not tracked in their
subprogram's retainedNodes, that crash LTO with Chromium.

This is meant to be committed along with https://reviews.llvm.org/D144006.

### CODE REVIEW

File: llvm/lib/Transforms/Utils/CloneFunction.cpp
Comment (adrian-prantl) [inline llvm/lib/Transforms/Utils/CloneFunction.cpp:241]: Can you add a sentence explaining why some subprograms won't be cloned? Is it because they are inlined and a copy already exists?

**Review by adrian-prantl — COMMENTED** [review]
I think this LGTM, but it would be good if someone else also took a look.

**Comment by dzhidzhoev** [comment]
@dwblaikie
Could you please take a look at this when you have a chance?

**Review by dwblaikie — APPROVED** [review]
Yeah, looks OK to me - sorry for the delay.

**Comment by jmorse** [comment]
Hi,

We're seeing a crash with this reproducer https://gist.github.com/jmorse/b0248c3c9f9195487ffd7c7431a8d15e

llc: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2338: virtual void llvm::DwarfDebug::endFunctionImpl(const llvm::MachineFunction *): Assertion `LScopes.getAbstractScopesList().size() == NumAbstractSubprograms && "getOrCreateAbstractScope() inserted an abstract subprogram scope"' failed.

I'd previously posted the reproducer on https://reviews.llvm.org/D144006#4656728 , however I'd anonymised the IR too much to the point where it was broken in unrelated ways. Revision 2 of the gist, as linked, should produce the crash. I suspect the extra lexical scopes reachable through the retained-nodes list also need to be explored when the LexicalScopes object gets constructed, to avoid scopes being added late and causing containers to invalidate iterators. (Which is what that assertion is there to detect).

**Comment by zmodem** [comment]
We're hitting the same assert (tracking in https://crbug.com/1518841) but don't have a good reproducer to share (it's during lto of a large binary). Jeremy's reproducer seems good though.

Can we revert this while it's being investigated, please?

**Comment by dcci** [comment]
Hitting the same problem on a bunch of services. I am going to revert this to unblock people, our repro is similar enough to the one Jeremy shared.

```
$ ./exe
clang -cc1 version 18.0.0git based upon LLVM 18.0.0git default target x86_64-redhat-linux-gnu
clang++: /home/davidino/llvm-project/llvm/include/llvm/Support/Casting.h:109: static bool llvm::isa_impl_cl<llvm::DILocalScope, const llvm::DIScope *>::doit(const From *) [To = llvm::DILocalScope, From = const llvm::DIScope *]: Assertion `Val && "isa<> used on a null pointer"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/davidino/llvm-build-upstream/bin/clang++ @/tmp/fbcc.4ddf7_1j/compiler.argsfile
1.      Running pass 'Function Pass Manager' on module 'buck-out/v2/gen/fbcode/349d8ab5ad6260b7/dsi/logger/configs/CacheClientFciRequestLoggerConfig/__logger__/__objects__/Logger.cpp.o'.
2.      Running pass 'X86 Assembly Printer' on function '@_ZN8facebook6logger27CacheClientFciRequestLogger9invokeLogIDnDnEEibT_PKciT0_'
 #0 0x00007f123a20a448 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/davidino/llvm-build-upstream/bin/../lib/libLLVMSupport.so.18git+0x20a448)
 #1 0x00007f123a207eee llvm::sys::RunSignalHandlers() (/home/davidino/llvm-build-upstream/bin/../lib/libLLVMSupport.so.18git+0x207eee)
 #2 0x00007f123a20ab18 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f1239854db0 __restore_rt (/lib64/libc.so.6+0x54db0)
 #4 0x00007f12398a365c __pthread_kill_implementation (/lib64/libc.so.6+0xa365c)
 #5 0x00007f1239854d06 gsignal (/lib64/libc.so.6+0x54d06)
 #6 0x00007f12398287f3 abort (/lib64/libc.so.6+0x287f3)
 #7 0x00007f123982871b _nl_load_domain.cold (/lib64/libc.so.6+0x2871b)
 #8 0x00007f123984dca6 (/lib64/libc.so.6+0x4dca6)
 #9 0x00007f123d3f54c3 getRetainedNodeScope(llvm::MDNode const*) DwarfDebug.cpp:0:0
#10 0x00007f123d3f686a llvm::DwarfDebug::endFunctionImpl(llvm::MachineFunction const*) (/home/davidino/llvm-build-upstream/bin/../lib/../lib/libLLVMAsmPrinter.so.18git+0x10086a)
#11 0x00007f123d3ce46f llvm::DebugHandlerBase::endFunction(llvm::MachineFunction const*) (/home/davidino/llvm-build-upstream/bin/../lib/../lib/libLLVMAsmPrinter.so.18git+0xd846f)
#12 0x00007f123d3a70fb llvm::AsmPrinter::emitFunctionBody() (/home/davidino/llvm-build-upstream/bin/../lib/../lib/libLLVMAsmPrinter.so.18git+0xb10fb)
#13 0x00007f1240c3db08 llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) X86AsmPrinter.cpp:0:0
#14 0x00007f123e1de48d llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/davidino/llvm-build-upstream/bin/../lib/libLLVMCodeGen.so.18git+0x5de48d)
#15 0x00007f123a8a0842 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/davidino/llvm-build-upstream/bin/../lib/libLLVMCore.so.18git+0x4a0842)
#16 0x00007f123a8a7542 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/davidino/llvm-build-upstream/bin/../lib/libLLVMCore.so.18git+0x4a7542)
#17 0x00007f123a8a0f56 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/davidino/llvm-build-upstream/bin/../lib/libLLVMCore.so.18git+0x4a0f56)
#18 0x00007f123cbd248a codegen(llvm::lto::Config const&, llvm::TargetMachine*, std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, unsigned int, llvm::Module&, llvm::ModuleSummaryIndex const&) LTOBackend.cpp:0:0
#19 0x00007f123cbd35cb llvm::lto::thinBackend(llvm::lto::Config const&, unsigned int, std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, llvm::Module&, llvm::ModuleSummaryIndex const&, llvm::DenseMap<llvm::StringRef, std::unordered_set<unsigned long, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<unsigned long>>, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, std::unordered_set<unsigned long, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<unsigned long>>>> const&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long, void>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*>> const&, llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>*, std::vector<unsigned char, std::allocator<unsigned char>> const&)::$_6::operator()(llvm::Module&, llvm::TargetMachine*, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>) const LTOBackend.cpp:0:0
#20 0x00007f123cbd3329 llvm::lto::thinBackend(llvm::lto::Config const&, unsigned int, std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, llvm::Module&, llvm::ModuleSummaryIndex const&, llvm::DenseMap<llvm::StringRef, std::unordered_set<unsigned long, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<unsigned long>>, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, std::unordered_set<unsigned long, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<unsigned long>>>> const&, llvm::DenseMap<unsigned long, llvm::GlobalValueSummary*, llvm::DenseMapInfo<unsigned long, void>, llvm::detail::DenseMapPair<unsigned long, llvm::GlobalValueSummary*>> const&, llvm::MapVector<llvm::StringRef, llvm::BitcodeModule, llvm::DenseMap<llvm::StringRef, unsigned int, llvm::DenseMapInfo<llvm::StringRef, void>, llvm::detail::DenseMapPair<llvm::StringRef, unsigned int>>, llvm::SmallVector<std::pair<llvm::StringRef, llvm::BitcodeModule>, 0u>>*, std::vector<unsigned char, std::allocator<unsigned char>> const&) (/home/davidino/llvm-build-upstream/bin/../lib/../lib/libLLVMLTO.so.18git+0x5a329)
#21 0x00007f123e8a1d0a clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/home/davidino/llvm-build-upstream/bin/../lib/libclangCodeGen.so.18git+0x2a1d0a)
#22 0x00007f123ed589dc clang::CodeGenAction::ExecuteAction() (/home/davidino/llvm-build-upstream/bin/../lib/libclangCodeGen.so.18git+0x7589dc)
#23 0x00007f123cd98aaf clang::FrontendAction::Execute() (/home/davidino/llvm-build-upstream/bin/../lib/libclangFrontend.so.18git+0x198aaf)
#24 0x00007f123ccfb21d clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/davidino/llvm-build-upstream/bin/../lib/libclangFrontend.so.18git+0xfb21d)
#25 0x00007f1240ff67fa clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/davidino/llvm-build-upstream/bin/../lib/libclangFrontendTool.so.18git+0x57fa)
#26 0x0000000000216d0f cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/davidino/llvm-build-upstream/bin/clang+++0x216d0f)
#27 0x00000000002131df ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#28 0x0000000000212497 clang_main(int, char**, llvm::ToolContext const&) (/home/davidino/llvm-build-upstream/bin/clang+++0x212497)
```

**Comment by dcci** [comment]
Reverted in https://github.com/llvm/llvm-project/commit/b6f922fbf5e983122271aa12acb33f6172046d4d

**Comment by dzhidzhoev** [comment]
> Hi,
> 
> We're seeing a crash with this commit and reproducer https://gist.github.com/jmorse/b0248c3c9f9195487ffd7c7431a8d15e
> 
> llc: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2338: virtual void llvm::DwarfDebug::endFunctionImpl(const llvm::MachineFunction *): Assertion `LScopes.getAbstractScopesList().size() == NumAbstractSubprograms && "getOrCreateAbstractScope() inserted an abstract subprogram scope"' failed.
> 
> I'd previously posted the reproducer on https://reviews.llvm.org/D144006#4656728 , however I'd anonymised the IR too much to the point where it was broken in unrelated ways. Revision 2 of the gist, as linked, should produce the crash. I suspect the extra lexical scopes reachable through the retained-nodes list also need to be explored when the LexicalScopes object gets constructed, to avoid scopes being added late and causing containers to invalidate iterators. (Which is what that assertion is there to detect).

Hello,

Sorry for bothering you. Could you please share the command you used as an interestingness test to reduce the crash? I've tried running llc/opt/ld.lld, but I haven't been able to reproduce it :(

**Comment by dzhidzhoev** [comment]
> Hmmm, that's unexpected -- I reverted the revert (tree here, contains one unrelated commit: https://github.com/jmorse/llvm-project/tree/reapply-localvars-patch) and rebuilt. The assertion-failure occurs just with `llc foobar.ll -o out.o -filetype=obj`, where foobar.ll is the contents of the gist file linked. For the avoidance of doubt, it's:
> 
> ```
> $ wget https://gist.githubusercontent.com/jmorse/b0248c3c9f9195487ffd7c7431a8d15e/raw/e57ca0417b77fe04f9551a0a69ce58f0db697527/gistfile1.txt
> $ mv gistfile1.txt foobar.ll
> $ md5sum foobar.ll
> d6c888d4e163545da596071fb5143377  foobar.ll
> ```
> 
> I don't think there's anything special about the build -- it's a CMAKE_BUILD_TYPE=Release build with LLVM_ENABLE_ASSERTIONS=On, built with clang-14 as shipped by ubuntu.

Thanks a lot! I've stubbornly tried to pass it through llvm-as before feeding it to llc.

**Comment by dzhidzhoev** [comment]
> Hi,
> 
> We're seeing a crash with this commit and reproducer https://gist.github.com/jmorse/b0248c3c9f9195487ffd7c7431a8d15e
> 
> llc: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2338: virtual void llvm::DwarfDebug::endFunctionImpl(const llvm::MachineFunction *): Assertion `LScopes.getAbstractScopesList().size() == NumAbstractSubprograms && "getOrCreateAbstractScope() inserted an abstract subprogram scope"' failed.
> 
> I'd previously posted the reproducer on https://reviews.llvm.org/D144006#4656728 , however I'd anonymised the IR too much to the point where it was broken in unrelated ways. Revision 2 of the gist, as linked, should produce the crash. I suspect the extra lexical scopes reachable through the retained-nodes list also need to be explored when the LexicalScopes object gets constructed, to avoid scopes being added late and causing containers to invalidate iterators. (Which is what that assertion is there to detect).

It seems that the debug info metadata was already malformed before getting into the AsmPrinter pass, probably during the LTO process. I ran into trouble trying to find the root cause of the problem, since the information provided by the reproducer is not sufficient to investigate it.
Could you please provide an IR dump of the earlier stage of compilation or any source code?

**Comment by jmorse** [comment]
Rats; we should be able to pin down where the problem originates from -- would you be able to describe what's malformed about the metadata, as it's not clear to us (sorry). We'll probably be able to llvm-reduce (or similar) around that description.

Unfortunately IIRC it comes from a reasonably large codebase that we can't share, possibly @zmodem @dcci have more accessible sources? (Short response sorry; currently up a mountain).

**Comment by jmorse** [comment]
...ah, actually is it malformed because there's a DICompositeType in the retainedNodes list for a DISubprogram? I remember that the verifier considered that illegal before your patch landed, but not after.

**Comment by dzhidzhoev** [comment]
> ...ah, actually is it malformed because there's a DICompositeType in the retainedNodes list for a DISubprogram? I remember that the verifier considered that illegal before your patch landed, but not after.

Having this commit applied, DICompositeTypes with the scope of DISubprogram must appear in the corresponding subprogram's retainedNodes list, so that's not the issue.
However, in the reproducer, DICompositeType node no !54 ("type4") has a scope value !55 ("plimsoles") but appears in the retainedNodes list of subprogram !49 ("crumpets").
!54 must have either scope of !49 or belong to the subprogram !55's retainedNodes list.
This discrepancy was likely to happen in CloneFunctionInto if LTO was used. However, I'm not sure about that.

**Comment by jmorse** [comment]
I see what you're saying about the metadata being incorrect; I feel like I've seen it before, but can't pin it down. For the record, all the builds where we saw this assertion were thin/full LTO.

I've kicked off a reduction of a large reproducer that I have to hand; I'm not immensely confident that it'll complete in a reasonable timescale (i.e.: days) as it's a large project being reduced. If any of the other reporters have anything smaller they can reduce, that'd assist significantly.

**Comment by dzhidzhoev** [comment]
> I see what you're saying about the metadata being incorrect; I feel like I've seen it before, but can't pin it down. For the record, all the builds where we saw this assertion were thin/full LTO.
> 
> I've kicked off a reduction of a large reproducer that I have to hand; I'm not immensely confident that it'll complete in a reasonable timescale (i.e.: days) as it's a large project being reduced. If any of the other reporters have anything smaller they can reduce, that'd assist significantly.

Thanks a lot!

**Comment by jmorse** [comment]
```

**Comment by jmorse** [comment]
Several decades later; I applied this check to the verifier and reduced around it:
```
diff --git a/llvm/lib/IR/Verifier.cpp b/llvm/lib/IR/Verifier.cpp
index b1cf81a3dbdc..9e2250b584b1 100644
--- a/llvm/lib/IR/Verifier.cpp
+++ b/llvm/lib/IR/Verifier.cpp
@@ -1421,6 +1421,9 @@ void Verifier::visitDISubprogram(const DISubprogram &N) {
               "invalid retained nodes, expected DILocalVariable, DILabel, "
               "DIImportedEntity or DIType",
               &N, Node, Op);
+      if (DIType *T = dyn_cast<DIType>(Op)) {
+        CheckDI(T->getScope() == &N, "wrong scope for thingy", T, T->getScope(), &N);
+      }
     }
   }
   CheckDI(!hasConflictingReferenceFlags(N.getFlags()),
```

Which produced these two IR files:
https://gist.github.com/jmorse/fc7d5479171b9943ae27d0f03cd9db5c
https://gist.github.com/jmorse/17040c19a096dd3780274f8e58d97b16
Which when linked with `llvm-link aa.ll bb.ll -o cc.ll -S` hit the verifier assertion, and have output IR which exhibits the problem you identified @dzhidzhoev , where the DICompositeType for `handle` ends up in one DISubprograms retainedNodes list but points its scope at a different DISubprogram. Seemingly this is something to do with the merging (or not merging) of distinct DISubprograms during the IR linker, unfortunately I've zero knowledge about that portion of LLVM.

**Comment by jmorse** [comment]
I think I've nailed down the source of this problem, and it's the matter to do with `LLVMContext::enableDebugTypeODRUniquing` that aeubanks mentioned on https://reviews.llvm.org/D144006. LLVM will unique DICompositeTypes via their ODR-name, which isn't always safe with function-local types. It's another circumstance where an ODR-violation that doesn't actually cause a linking error can instead cause debug-info to be corrupt. I don't have a reproducer that forces the crash found here but I can demonstrate the fundamental problem.

To reproduce, take the rebased patch from the tip of https://github.com/jmorse/llvm-project/tree/rebased-composite-type-fix-4 , and this code from llvm/test/Linker/odr-lambda-1.ll:

    class Error {};
    template <typename HandlerTs>
    void handleAllErrors(HandlerTs Handlers) {}
    inline void consumeError(Error Err) {
      handleAllErrors( []() {});
    }
    void ArchiveMemberHeader()
    {
      consumeError(Error());
    }

Compile thusly:

    clang input.cpp -o out1.ll -emit-llvm -S -c  -g
    clang input.cpp -o out2.ll -emit-llvm -S -c  -g
    vi out1.ll # Delete all LLVM function definitions to avoid linking-errors
    llvm-link out1.ll out2.ll -o out3.ll -S

The link should succeed because there are no symbol conflicts. In the output IR in [out3.ll.txt](https://github.com/user-attachments/files/16354671/out3.ll.txt), observe that there are:
 * Two DISubprogram records for the "consumeError" function,
 * One DICompositeType for `_ZTSZ12consumeError5ErrorEUlvE_`, "`typeinfo name for consumeError(Error)::{lambda()#1}`"
 * The DICompositeType is the retainedNodes member of one DISubprogram, but...
 * The DICompositeType's Scope pointer is to the _other_ DISubprogram, that is otherwise unlinked to the rest of the metadata hierarchy.

It's this mismatch where the scope of an element in the retained nodes points somewhere unexpected, which causes the assertion failure reported. It's all because of the ODR-uniquing by type name that LLVM can do, a rationale is here:

https://github.com/llvm/llvm-project/blob/2604830aacdd563715da030d0396b565e912436f/clang/lib/CodeGen/CGDebugInfo.cpp#L1086

When DICompositeType::buildODRType de-duplicates on the basis of the type name, if it's a type inside a function then the scope pointer will point at the DISubprogram where the type was first seen. If there's an ODR violation, this might not be the same as the DISubprogram that is eventually selected for that function.

IIRC there's precedent in LLVM for throwing away lots of debug-info in the presence of ODR violations, but we should avoid hard errors.

**Comment by rnk** [comment]
I'm not seeing how your example constitutes an ODR violation, or how merging these lambda types by mangled name is incorrect. They are equivalent in your example. It seems like the issue has more to do with the details of how exactly we do the merge, and where the metadata references from DICompositeType to DISubprogram point.

**Comment by dexonsmith** [comment]
> An immediate fix would be to not set the "identifier" field for the DICompositeType when it's created if it's inside a function scope to avoid ODRUniqing. I've only got a light understanding of what the identifier field is for, so there might be unexpected consequences, plus there'll be a metadata/DWARF size cost to that.

Actually, that seems like probably the correct fix. The identifier field is precisely for "ODR-unique-based-on-this" and only for that, IIRC (@aprantl, can you confirm? these are old memories for me at this point...).

If the function-local types should not be ODR-uniqued, then dropping the identifier field sounds correct.

**Comment by rnk** [comment]
> If the function-local types should not be ODR-uniqued, then dropping the identifier field sounds correct.

I can't speak to the complexities of the alternative, but I'm immediately suspicious of this direction. We have stable manglings for static locals in inline functions and static data members of local classes and static locals in methods in local classes in inline functions... Surely if we can compute a unique mangled name for this local class, we should use it as the identifier and merge it. It seems important that, for size reasons, we figure out a way to do this for lambdas, which are common in inline functions in headers.

**Comment by jmorse** [comment]
Hadn't realised lambda types will pop up everywhere, the cost might be too high. I'll test with building clang and our internal codebases to see if the fix works and what the cost is.

> Surely if we can compute a unique mangled name for this local class, we should use it as the identifier and merge it.

Indeed, it's not the ODR-uniquing per-se that's causing the trouble here, it's the fact that we do not merge together the DISubprograms that are the containing scope for that type upon module linking. I'm not familiar with the design decisions behind that, but a courtesy glance suggests it's important to identify the compilation unit that a DISubprogram instance came from. We're then stuck with multiple DISubprograms and a single DIComposite type referring to one of them as being its scope, which presumably breaks a lot of assumptions.

Perhaps a worthwhile direction is separating DISubprograms into abstract and definition portions, much like DWARF does, which would allow us to put these types into the declaration and have a well defined scope hierarchy: types that get uniqued refer to the abstract-DISubprogram for their lexical scope. Perhaps that's can be achieved with the existing distinction between definition DISubprograms and "declaration" DISubprograms? A quick survey of callers to `DISubprogram::getDefinition` and `isDefinition` indicates there aren't many decisions made throughout LLVM based on the definition/declaration split, and there are comments in LLVMContextImpl.h suggesting declaration-DISubprograms are used in ODR'd types too.

If we took that route, the scope-chain would still become a tree of scopes instead of distinct chains, and we would still need to fix the problem of CodeGen LexicalScopes encountering the abstract/declaration DISubprogram. However, I think we'd have the reasonable solution of mapping declaration-DISubprograms in a function to their definition-DISubprogram (assuming the definition had been found already), which is much more reliable than seeking a DISubprogram by name or something.

**Comment by rnk** [comment]
> Perhaps a worthwhile direction is separating DISubprograms into abstract and definition portions, much like DWARF does, which would allow us to put these types into the declaration and have a well defined scope hierarchy: types that get uniqued refer to the abstract-DISubprogram for their lexical scope. 

This sounds like a promising direction to me, but I have no idea what the relative costs are.

**Comment by jmorse** [comment]
I've placed a re-application of this code in #119001, along with a commit that attempts to fix the matter discovered here.  To avoid multiple DICompositeTypes in multiple DISubprograms being ODRUnique'd, we instead put them in the declaration DISubprogram which should be unique'd anyway.

**Pull request merged**
